<a href="https://colab.research.google.com/github/tanishqacodes/DATA-SCIENCE-MASTERS/blob/main/19_MAR_ASS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###Q1. What is Min-Max scaling, and how is it used in data preprocessing? Provide an example to illustrate its application.

Min-Max scaling, also known as normalization, is a data preprocessing technique used to transform numerical features within a specific range. It rescales the data, bringing it to a common scale, typically between 0 and 1.

The purpose of Min-Max scaling is to eliminate the influence of the original values' magnitude, preventing certain features from dominating the learning algorithms simply due to their larger values.


**X_scaled = (X - X_min) / (X_max - X_min)**

EXAMPLE :-

In [1]:
from sklearn.preprocessing import MinMaxScaler
min_max=MinMaxScaler()

In [2]:
import seaborn as sns
df=sns.load_dataset('taxis')
df.head()

pickup             dropoff  passengers  distance  fare   tip  \
0 2019-03-23 20:21:09 2019-03-23 20:27:24           1      1.60   7.0  2.15   
1 2019-03-04 16:11:55 2019-03-04 16:19:00           1      0.79   5.0  0.00   
2 2019-03-27 17:53:01 2019-03-27 18:00:25           1      1.37   7.5  2.36   
3 2019-03-10 01:23:59 2019-03-10 01:49:51           1      7.70  27.0  6.15   
4 2019-03-30 13:27:42 2019-03-30 13:37:14           3      2.16   9.0  1.10   

   tolls  total   color      payment            pickup_zone  \
0    0.0  12.95  yellow  credit card        Lenox Hill West   
1    0.0   9.30  yellow         cash  Upper West Side South   
2    0.0  14.16  yellow  credit card          Alphabet City   
3    0.0  36.95  yellow  credit card              Hudson Sq   
4    0.0  13.40  yellow  credit card           Midtown East   

            dropoff_zone pickup_borough dropoff_borough  
0    UN/Turtle Bay South      Manhattan       Manhattan  
1  Upper West Side South      Manhattan       Manhattan  
2           West Village      Manhattan       Manhattan  
3         Yorkville West      Manhattan       Manhattan  
4         Yorkville West      Manhattan       Manhattan

In [4]:
min_max.fit_transform(df[['distance','fare','tip']])

array([[0.04359673, 0.04026846, 0.06475904],
       [0.02152589, 0.02684564, 0.        ],
       [0.0373297 , 0.04362416, 0.07108434],
       ...,
       [0.11280654, 0.10067114, 0.        ],
       [0.03051771, 0.03355705, 0.        ],
       [0.10490463, 0.09395973, 0.10120482]])

### Q2. What is the Unit Vector technique in feature scaling, and how does it differ from Min-Max scaling? Provide an example to illustrate its application.

The Unit Vector technique, also known as normalization, is another data preprocessing method used for feature scaling. Unlike Min-Max scaling, which scales the data to a specific range, Unit Vector scaling aims to normalize the vector representation of each data point.

The Unit Vector technique scales each data point such that its Euclidean norm (or length) becomes 1. It ensures that the vector retains its direction while adjusting its magnitude to a unit length. This scaling technique is particularly useful when the direction of the data points is important, such as in cases where similarity or distance calculations are involved.

**X_scaled = X / ||X||**



In [5]:
from sklearn.preprocessing import normalize

In [7]:
import pandas as pd
pd.DataFrame(normalize(df[['distance','fare','tip']]))

0         1         2
0     0.213461  0.933894  0.286839
1     0.156064  0.987747  0.000000
2     0.171657  0.939731  0.295702
3     0.267899  0.939386  0.213971
4     0.231742  0.965592  0.118017
...        ...       ...       ...
6428  0.160133  0.960800  0.226322
6429  0.307453  0.951563  0.000000
6430  0.250500  0.968117  0.000000
6431  0.183497  0.983020  0.000000
6432  0.242956  0.946580  0.212034

[6433 rows x 3 columns]

###Q3. What is PCA (Principle Component Analysis), and how is it used in dimensionality reduction? Provide an example to illustrate its application.

Principal Component Analysis (PCA) is a dimensionality reduction technique used to transform a high-dimensional dataset into a lower-dimensional space. It achieves this by identifying the principal components, which are orthogonal directions that capture the maximum variance in the data. These principal components represent new uncorrelated variables that can be used to represent the original dataset with reduced dimensions.

PCA works by linearly transforming the data onto a new coordinate system defined by the principal components. The first principal component corresponds to the direction of maximum variance in the data, and each subsequent component captures as much remaining variance as possible while being orthogonal (uncorrelated) to the previous components. This allows PCA to capture the most important patterns and structures in the data.

Here's an example to illustrate the application of PCA:

Suppose we have a dataset with three numerical features: "Height," "Weight," and "Age." Each data point represents an individual.

Original dataset:

|Height (cm) | Weight (kg) | Age (years) |
|:-----------|:------------|:------------|
| 170	       | 60	         | 30          |
| 160	       | 55          | 25          |
| 180        | 70	         | 35          |
| 165        | 63          | 28          |
| 175        | 68          | 32          |

To apply PCA for dimensionality reduction, we first standardize the data by subtracting the mean and scaling to unit variance. This step ensures that all features contribute equally to the analysis.

Next, we compute the covariance matrix or the correlation matrix of the standardized data. The covariance matrix represents the relationships between the features and their variances.

From the covariance matrix, we calculate the eigenvectors and eigenvalues. The eigenvectors represent the principal components, and the eigenvalues indicate the amount of variance explained by each component.

Suppose the eigenvectors and eigenvalues are as follows:

Eigenvectors:

|PC1	|PC2	|PC3|
|-----|-----|----|
|0.58	|0.69	|0.44|
|0.67	|-0.33	|-0.66|
|-0.47|	0.67	|-0.56|

Eigenvalues: [1.92, 0.64, 0.04]

The eigenvectors are sorted based on their corresponding eigenvalues in descending order. This indicates the amount of variance explained by each principal component.

Now, we can choose the desired number of principal components to retain. For instance, if we want to reduce the dimensionality to two, we select the top two eigenvectors (PC1 and PC2).

We project the original data onto the selected principal components to obtain the new lower-dimensional representation. This can be done by multiplying the standardized data by the selected eigenvectors.

Projected dataset (using PC1 and PC2):

|PC1|	PC2|
|---|----|
|-0.36|	-1.32|
|-0.62|	1.47|
|1.78	|-0.18|
|-0.41|	-0.71|
|-0.38|	0.74|

The projected dataset represents the original data in a reduced two-dimensional space, where each data point is now represented by its corresponding values along the principal components.

PCA allows us to reduce the dimensionality of the dataset while preserving as much information (variance) as possible. The reduced-dimensional representation can be useful for visualization, exploration, and subsequent analysis tasks such as clustering or classification.

### Q4. What is the relationship between PCA and Feature Extraction, and how can PCA be used for Feature Extraction? Provide an example to illustrate this concept.

PCA can be used for feature extraction, which involves transforming the original features into a new set of features that are more informative or representative of the underlying patterns in the data. In this context, PCA identifies the most important features or combinations of features, known as principal components, and represents the data using these components.

Feature extraction using PCA involves the following steps:

Standardize the data: Start by standardizing the dataset to have zero mean and unit variance. This step ensures that all features contribute equally to the PCA analysis.

Compute the covariance matrix or correlation matrix: Calculate the covariance matrix (or correlation matrix) of the standardized data. The covariance matrix represents the relationships between the features and their variances.

Compute eigenvectors and eigenvalues: Determine the eigenvectors and eigenvalues of the covariance matrix. The eigenvectors represent the principal components, while the eigenvalues indicate the amount of variance explained by each component.

Select principal components: Choose the desired number of principal components to retain based on the eigenvalues. Retaining a subset of the principal components allows for dimensionality reduction and feature extraction.

Project the data onto the selected principal components: Multiply the standardized data by the selected eigenvectors (principal components) to obtain the new feature representations.

### Q5. You are working on a project to build a recommendation system for a food delivery service. The dataset contains features such as price, rating, and delivery time. Explain how you would use Min-Max scaling to preprocess the data.

To preprocess the data for building a recommendation system for a food delivery service, you can use Min-Max scaling to ensure that the features are on a similar scale. Here's how you can apply Min-Max scaling to the dataset:

1. Identify the features: Review the dataset and identify the relevant features that you want to include in the recommendation system. In this case, the features would be price, rating, and delivery time.

2. Find the minimum and maximum values: Calculate the minimum and maximum values for each feature. For example, determine the minimum and maximum prices, ratings, and delivery times in the dataset.

3. Apply Min-Max scaling formula: Use the Min-Max scaling formula to scale the values of each feature. The formula is:

**X_scaled = (X - X_min) / (X_max - X_min)**

4. Scale the values: Apply the Min-Max scaling formula to each value of the features separately. For example, for price, you would subtract the minimum price and divide by the range (maximum price minus minimum price). Repeat this process for rating and delivery time as well.

5. Obtain the scaled dataset: After scaling all the values, you will have a new dataset where each feature is transformed to a range between 0 and 1. This ensures that all features are on a similar scale and eliminates the influence of their original magnitudes.

6. By using Min-Max scaling, the recommendation system can effectively handle features with different value ranges. This preprocessing step ensures that no single feature dominates the recommendation process based solely on its larger or smaller values. Instead, all features are given equal importance in determining the recommendations.

### Q6. You are working on a project to build a model to predict stock prices. The dataset contains many features, such as company financial data and market trends. Explain how you would use PCA to reduce the dimensionality of the dataset.

To reduce the dimensionality of the dataset in a project focused on predicting stock prices, you can use PCA (Principal Component Analysis) as a dimensionality reduction technique. Here's how you can use PCA in this scenario:

1. Identify the relevant features: Review the dataset and identify the features that are relevant for predicting stock prices. These features may include company financial data (e.g., revenue, earnings, debt) and market trends (e.g., interest rates, market indices).

2. Standardize the data: Before applying PCA, it's important to standardize the features to have zero mean and unit variance. Standardization ensures that all features contribute equally to the PCA analysis, as the algorithm is sensitive to the relative scales of the features.

3. Apply PCA: Once the data is standardized, apply PCA to the dataset. PCA will identify the principal components, which are new uncorrelated variables that capture the maximum variance in the data.

4. Determine the number of principal components: Decide on the number of principal components to retain. This decision can be based on the cumulative explained variance or domain knowledge. Retaining fewer components will result in dimensionality reduction while still preserving the most important patterns in the data.

5. Project the data onto the selected principal components: Transform the dataset by projecting the original features onto the selected principal components. This step will create a new dataset with reduced dimensions, where each observation is represented by the values along the principal components.

6. Use the reduced dataset for modeling: The reduced-dimensional dataset obtained from PCA can be used as input for modeling algorithms to predict stock prices. The reduced dimensions may help in reducing noise, focusing on the most relevant information, and improving model performance.

Using PCA for dimensionality reduction in the context of predicting stock prices can be beneficial as it reduces the computational complexity and helps to handle multicollinearity among the features. By capturing the most important patterns in the data, PCA allows you to focus on the most informative components while discarding less relevant ones. However, it's important to note that the interpretability of the reduced features may be reduced compared to the original features.

### Q7. For a dataset containing the following values: [1, 5, 10, 15, 20], perform Min-Max scaling to transform the values to a range of -1 to 1.

To perform Min-Max scaling on the given dataset, which is [1, 5, 10, 15, 20], and transform the values to a range of -1 to 1, follow these steps:

Find the minimum and maximum values of the dataset:

Minimum value (X_min): 1
Maximum value (X_max): 20
Apply the Min-Max scaling formula:

`X_scaled = (X - X_min) / (X_max - X_min)`

Scale each value in the dataset using the formula:

For the value 1: X_scaled = (`1 - 1) / (20 - 1) = 0 / 19 = 0`

For the value 5: X_scaled = `(5 - 1) / (20 - 1) = 4 / 19 ≈ 0.211`

For the value 10: X_scaled = `(10 - 1) / (20 - 1) = 9 / 19 ≈ 0.474`

For the value 15: X_scaled = `(15 - 1) / (20 - 1) = 14 / 19 ≈ 0.737`





For the value 20: X_scaled = (20 - 1) / (20 - 1) = 19 / 19 = 1

Obtain the scaled dataset:

The scaled dataset using Min-Max scaling and transforming the values to a range of -1 to 1 is:
[0, 0.211, 0.474, 0.737, 1]

The original values in the dataset [1, 5, 10, 15, 20] have been transformed to a new range between -1 and 1 using Min-Max scaling. This scaling technique ensures that all values are proportionally scaled based on the range of the original data.

### Q8. For a dataset containing the following features: [height, weight, age, gender, blood pressure], perform Feature Extraction using PCA. How many principal components would you choose to retain, and why?

To perform feature extraction using PCA on the given dataset with features [height, weight, age, gender, blood pressure], follow these steps to determine the number of principal components to retain:

1. Prepare the dataset: Ensure the dataset is properly preprocessed, which may involve handling missing values, encoding categorical variables (such as gender), and scaling the numerical features.

2. Standardize the data: Standardize the numerical features (height, weight, age, and blood pressure) to have zero mean and unit variance. This step ensures that all features contribute equally to the PCA analysis.

3. Apply PCA: Perform PCA on the standardized dataset. This will yield the principal components, eigenvectors, and eigenvalues.

4. Analyze the explained variance: Examine the eigenvalues or the explained variance ratio associated with each principal component. The explained variance ratio indicates the proportion of variance in the data that is explained by each principal component. The cumulative explained variance can help determine the number of principal components to retain.

5. Determine the number of principal components: Decide on the number of principal components to retain based on the cumulative explained variance. Retaining a sufficient number of principal components is crucial to capture a significant portion of the variance in the data while reducing dimensionality.

The specific number of principal components to retain can vary depending on the dataset and the desired trade-off between dimensionality reduction and preserving information. Common approaches include:

* Retaining a fixed number of principal components: You may choose to retain a specific number of principal components based on prior knowledge or the desired level of dimensionality reduction. For example, retaining the top two or three principal components.

* Retaining principal components with a minimum explained variance threshold: You can set a minimum threshold for the cumulative explained variance (e.g., 90% or 95%) and retain principal components until that threshold is reached.

* Visual analysis: Plotting the scree plot, which shows the eigenvalues or explained variance ratio, can provide insights into the "elbow point" where the eigenvalues start to level off. This point may indicate a suitable number of principal components to retain.

It's challenging to determine the exact number of principal components to retain without knowing the specific dataset and its characteristics. It is recommended to experiment with different options, such as the explained variance ratio and visual analysis, to find an appropriate balance between dimensionality reduction and information preservation.